diff -ru connman-1.1/src/connman.service.in connman-1.1-tracing/src/connman.service.in
--- connman-1.1/src/connman.service.in	2010-11-25 11:08:50.000000000 +0000
+++ connman-1.1-tracing/src/connman.service.in	2012-07-02 06:41:03.419616545 +0000
@@ -5,7 +5,8 @@
 [Service]
 Type=dbus
 BusName=net.connman
-ExecStart=@prefix@/sbin/connmand -n
+EnvironmentFile=-/etc/tracing/connman
+ExecStart=@prefix@/sbin/connmand -n $TRACING
 
 [Install]
 WantedBy=multi-user.target
